### PR TITLE
Add support in AWS Batch Operator for multinode jobs

### LINKED ISSUE #25522: Support AWS Batch multinode job types
### Description

Support [multinode job types](https://docs.aws.amazon.com/batch/latest/userguide/multi-node-parallel-jobs.html) in the [AWS Batch Operator](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/operators/batch.py). 

The [boto3 `submit_job` method](https://boto3.amazonaws.com/v1/documentation/api/1.9.88/reference/services/batch.html#Batch.Client.submit_job) supports container, multinode, and array batch jobs with the mutually exclusive `nodeOverrides` and `containerOverrides` (+ `arrayProperties`) parameters. But currently the AWS Batch Operator only supports submission of container jobs and array jobs by hardcoding the boto3 `submit_job` parameter `containerOverrides`: https://github.com/apache/airflow/blob/3c08cefdfd2e2636a714bb835902f0cb34225563/airflow/providers/amazon/aws/operators/batch.py#L200 & https://github.com/apache/airflow/blob/3c08cefdfd2e2636a714bb835902f0cb34225563/airflow/providers/amazon/aws/hooks/batch_client.py#L99

The [`get_job_awslogs_info`](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/batch_client.py#L419) method in the batch client hook is also hardcoded for the container type job: https://github.com/apache/airflow/blob/3c08cefdfd2e2636a714bb835902f0cb34225563/airflow/providers/amazon/aws/hooks/batch_client.py#L425

To support multinode jobs the `get_job_awslogs_info` method would need to access `nodeProperties` from the [`describe_jobs`](https://boto3.amazonaws.com/v1/documentation/api/1.9.88/reference/services/batch.html#Batch.Client.describe_jobs) response.

### Use case/motivation

Multinode jobs are a supported job type of AWS Batch, are supported by the underlying boto3 library, and should be also be available to be managed by Airflow. I've extended the AWS Batch Operator for our own use cases, but would prefer to not maintain a separate operator.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by camilleanne**
* Adds support for AWS Batch multinode jobs by allowing a `node_overrides` json object to be passed through to the [boto3 submit_job](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/batch.html#Batch.Client.submit_job) method.
* Adds support for multinode jobs by properly parsing the output of `describe_jobs` (which is different for container vs multinode) to extract the log stream name.

closes: #25522

I had a hard time running tests locally, so I'm opening as a draft PR initially although I don't anticipate any changes beyond syncing with main, but I'd like to confirm test success before making available for review.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/operators/batch.py [hunk 1]
```diff
@@ -151,8 +155,9 @@ def __init__(
         self.job_name = job_name
         self.job_definition = job_definition
         self.job_queue = job_queue
-        self.overrides = overrides or {}
-        self.array_properties = array_properties or {}
+        self.container_overrides = overrides
+        self.node_overrides = node_overrides
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/operators/batch.py]: Just a question. Did you check are this arguments mutually exclusive? 
[AWS API](https://docs.aws.amazon.com/batch/latest/APIReference/API_SubmitJob.html) doesn't mention it however everything might possible because even new [eksPropertiesOverride](https://docs.aws.amazon.com/batch/latest/APIReference/API_SubmitJob.html#API_SubmitJob_RequestSyntax) not marked as exclusive.

Comment (vandonr-amz) [reply airflow/providers/amazon/aws/operators/batch.py]: yes they are, specifying both returns an input validation error.
BTW @Taragolis , could you approve running the workflow to help get this PR ready-to-review ? 🙏

**Comment by camilleanne** [comment]
Ok I got a handle on all tests finally :) ready for review now.

File: airflow/providers/amazon/aws/hooks/batch_client.py [hunk 2]
```diff
@@ -416,8 +416,43 @@ def get_job_awslogs_info(self, job_id: str) -> dict[str, str] | None:
...
+
+        if job_node_properties:
+            job_node_range_properties = job_node_properties.get("nodeRangeProperties", {})
+            if len(job_node_range_properties) > 1:
+                self.log.warning(
+                    "AWS Batch job (%s) has more than one node group. Only returning logs from first group.",
+                    job_id,
+                )
+            log_configuration = (
+                job_node_range_properties[0].get("container", {}).get("logConfiguration", {})
```
Comment (vandonr-amz) [inline airflow/providers/amazon/aws/hooks/batch_client.py:435]: Is it possible to have zero element in the array ? i.e. should we add a check on `len == 0` and a user-friendly error message ?

File: airflow/providers/amazon/aws/hooks/batch_client.py [hunk 3]
```diff
@@ -416,8 +416,43 @@ def get_job_awslogs_info(self, job_id: str) -> dict[str, str] | None:
...
+                awslogs_stream_name = job_attempts[-1].get("container", {}).get("logStreamName")
+            else:
+                awslogs_stream_name = None
+
+        elif job_container_desc:
+            log_configuration = job_container_desc.get("logConfiguration", {})
+            awslogs_stream_name = job_container_desc.get("logStreamName")
+        else:
+            self.log.warning(
+                "AWS Batch job (%s) is neither a container nor multinode job. Log info not found."
```
Comment (vandonr-amz) [inline airflow/providers/amazon/aws/hooks/batch_client.py]: Maybe this could be an error log, considering the user-provided input is invalid for this kind of request ? 

The other warning logs in this method are mostly informative (there are several node groups, which is important info for the user to know, but doesn't require any action), this one I think requires user action, and thus more attention.

File: tests/providers/amazon/aws/hooks/test_batch_client.py [hunk 4]
```diff
@@ -309,6 +326,40 @@ def test_job_splunk_logs(self, caplog):
...
+                                            "awslogs-region": AWS_REGION,
+                                        },
+                                    }
+                                },
+                            }
+                        ],
+                    },
+                }
+            ]
+        }
```
Comment (vandonr-amz) [inline tests/providers/amazon/aws/hooks/test_batch_client.py:359]: beautiful 😄

File: tests/providers/amazon/aws/operators/test_batch.py [hunk 5]
```diff
@@ -187,6 +188,47 @@ def test_kill_job(self):
         self.client_mock.terminate_job.assert_called_once_with(jobId=JOB_ID, reason="Task killed by the user")
 
 
+class TestBatchOperatorTrimmedArgs:
+    """test class that does not inherit from unittest.TestCase"""
```
Comment (vandonr-amz) [inline tests/providers/amazon/aws/operators/test_batch.py]: I added this class before @Taragolis removed `unittest.TestCase` in all AWS tests, this can be merged with the class above now :)

File: airflow/providers/amazon/aws/operators/batch.py [hunk 6]
```diff
@@ -108,12 +110,13 @@ class BatchOperator(BaseOperator):
         "job_queue",
         "overrides",
```
Comment (vandonr-amz) [inline airflow/providers/amazon/aws/operators/batch.py]: TIL the names here need to match not the parameter name in the ctor but the name of the attribute in the class, so:
```suggestion
        "container_overrides",
```

Thinking about it, I wonder if this is a breaking change... I don't know too well how this works 😬

Comment (Taragolis) [reply airflow/providers/amazon/aws/operators/batch.py]: @camilleanne @vandonr-amz I recommend deprecate old parameter but keep it for a while, so users would have a time for change their code.
It is much easier achieve in subclass of `BaseOperator` (this PR case), because all arguments are keyword. 

Some examples

**Use old attribute value only if new attribute not set**
https://github.com/apache/airflow/blob/11f30a887c77f9636e88e31dffd969056132ae8c/airflow/providers/slack/operators/slack_webhook.py#L95-L104

**Use old attribute value only if it equal new attribute value or not new attribute not set**
https://github.com/apache/airflow/blob/11f30a887c77f9636e88e31dffd969056132ae8c/airflow/providers/amazon/aws/operators/athena.py#L91-L101

Comment (vandonr-amz) [reply airflow/providers/amazon/aws/operators/batch.py]: given the fact that it was just a rename for readability, I wonder if it wouldn't be simpler to just keep the old name ?
but then it could be confusing to users...

Comment (Taragolis) [reply airflow/providers/amazon/aws/operators/batch.py]: I thought it fine if rename this argument because one day we might also add `eks_properties_overrides`, so `override` might confuse users more rather than change attribute name.
And deprecation warning give time to change arguments in end users DAG code.

Comment (ferruzzi) [reply airflow/providers/amazon/aws/operators/batch.py]: > TIL the names here need to match not the parameter name in the ctor but the name of the attribute in the class

To translate that into ELI5: if you have `self.foo = bar` in your operator, the template field would be named "foo" to match "self.foo", not "bar". :P

**Comment by o-nikolas** [comment]
Hey @camilleanne,

Any plans to pick this one up again?

**Comment by o-nikolas** [comment]
@vandonr-amz is taking on this work on in #29522

Closing this PR with that context.
